crates/noirc_evaluator/src/ssa_refactor/ir/value_map.rs

```diff
@@ -0,0 +1,106 @@
+use std::collections::HashMap;
+
+use iter_extended::vecmap;
+
+use super::{
+    basic_block::BasicBlockId,
+    dfg::{DataFlowGraph, InsertInstructionResult},
+    instruction::{Instruction, InstructionId},
+    types::Type,
+    value::ValueId,
+};
+
+/// Maps old value ids to a new set of value ids, not necessarily in the same function.
+/// This is useful for optimization passes which rewrite instructions and values.
+/// Notably `map_instruction` and `push_instruction` are useful operations for such
+/// passes so that they need to worry less about remembering to update each ValueId
+/// within each Instruction.
+#[derive(Default)]
+pub(crate) struct ValueMap {
+    map: HashMap<ValueId, ValueId>,
+}
+
+impl ValueMap {
+    /// Insert a new mapping into this value map, overwriting any previous entry if present
+    pub(crate) fn insert(&mut self, old_id: ValueId, new_id: ValueId) {
+        self.map.insert(old_id, new_id);
+    }
+
+    /// Insert the given old_id -> new_id mapping if old_id isn't already in the map
+    pub(crate) fn insert_if_not_present(&mut self, old_id: ValueId, new_id: ValueId) {
+        self.map.entry(old_id).or_insert(new_id);
+    }
+
+    /// Retrieve the mapped value or return None otherwise
+    pub(crate) fn get(&self, value: ValueId) -> Option<ValueId> {
```

Comment:
The distinction between `get` and `get_value` isn't intuitive from the name. (Unless this is a rust convention I'm not familiar with?) Perhaps `get_substitution` might be clearer that `get`.